python/pylibvw.cc

```diff
@@ -57,6 +58,117 @@ const size_t pPDF = 10;
+
+public:
+  OptionManager(VW::config::options_i& options, std::vector<std::string>& enabled_reductions, py::object py_class)
+      : m_opt(options)
+      , m_enabled_reductions(enabled_reductions)
+      , m_option_group_dic(options.get_collection_of_options())
+      , m_py_opt_class(py_class)
+  {
+    default_group_name = static_cast<VW::config::options_boost_po*>(&options)->m_default_tint;
+  }
+
+  template <typename T>
+  py::object* value_to_pyobject(VW::config::typed_option<T>& opt)
+  {
+    return new py::object(
+        m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary, opt.m_allow_override,
+            opt.value(), m_opt.was_supplied(opt.m_name), opt.default_value(), opt.default_value_supplied()));
+  }
+
+  template <typename T>
+  py::object* value_to_pyobject(VW::config::typed_option<std::vector<T>>& opt)
+  {
+    py::list values;
+
+    auto vec = opt.value();
+    if (vec.size() > 0)
+    {
+      for (auto const& opt : vec) { values.append(py::object(opt)); }
+    }
+
+    return new py::object(m_py_opt_class(opt.m_name, opt.m_help, opt.m_short_name, opt.m_keep, opt.m_necessary,
+        opt.m_allow_override, values, m_opt.was_supplied(opt.m_name), py::list(), opt.default_value_supplied()));
+  }
+
+  template <typename T>
+  py::object* transform_if_t(VW::config::base_option& base_option)
+  {
+    if (base_option.m_type_hash == typeid(T).hash_code())
+    {
+      auto typed = dynamic_cast<VW::config::typed_option<T>&>(base_option);
+      return value_to_pyobject(typed);
+    }
+
+    return nullptr;
+  }
+
+  template <typename TTypes>
+  py::object base_option_to_pyobject(VW::config::base_option& options)
+  {
+    py::object* temp = transform_if_t<typename TTypes::head>(options);
+    if (temp != nullptr)
+    {
+      auto repack = py::object(*temp);
+      delete temp;
+      return repack;
+    }
+
+    return base_option_to_pyobject<typename TTypes::tail>(options);
+  }
+
+  py::object get_vw_option_pyobjects(bool enabled_only)
+  {
+    py::dict dres;
+    auto it = m_option_group_dic.begin();
+
+    while (it != m_option_group_dic.end())
+    {
+      auto reduction_enabled =
+          std::find(m_enabled_reductions.begin(), m_enabled_reductions.end(), it->first) != m_enabled_reductions.end();
+
+      if (((it->first).compare(default_group_name) != 0) && enabled_only && !reduction_enabled)
+      {
+        it++;
+        continue;
+      }
+
+      py::list option_groups;
+
+      for (auto options_group : it->second)
+      {
+        py::list options;
+        for (auto opt : options_group.m_options)
+        {
+          auto temp = base_option_to_pyobject<VW::config::supported_options_types>(*opt.get());
+          options.append(temp);
+        }
+
+        option_groups.append(py::make_tuple(options_group.m_name, options));
+      }
+
+      dres[it->first] = option_groups;
+
+      it++;
+    }
+    return dres;
+  }
+};
+
+template <>
```

Comment:
```suggestion
// specialization needed to compile, this should never be reached since we always use VW::config::supported_options_types
template <>
```